ext/drreg/drreg.c

```diff
@@ -1609,20 +1609,26 @@ is_our_spill_or_restore(void *drcontext, instr_t *instr, bool *spill OUT,
                 opnd_get_disp(dr_reg_spill_slot_opnd(drcontext, SPILL_SLOT_1));
             uint DR_max_offs = opnd_get_disp(
                 dr_reg_spill_slot_opnd(drcontext, dr_max_opnd_accessible_spill_slot()));
+            uint max_DR_slot = (uint)dr_max_opnd_accessible_spill_slot();
             if (DR_min_offs > DR_max_offs) {
                 if (offs > DR_min_offs) {
                     slot = (offs - DR_min_offs) / sizeof(reg_t);
+                } else if (offs < DR_max_offs) {
+                    /* Fix hidden slot regardless of low-to-high or vice versa. */
+                    slot = max_DR_slot + 1;
                 } else {
                     slot = (DR_min_offs - offs) / sizeof(reg_t);
                 }
             } else {
                 if (offs > DR_max_offs) {
                     slot = (offs - DR_max_offs) / sizeof(reg_t);
+                } else if (offs < DR_min_offs) {
+                    /* Fix hidden slot regardless of low-to-high or vice versa. */
+                    slot = max_DR_slot + 1;
                 } else {
                     slot = (offs - DR_min_offs) / sizeof(reg_t);
                 }
             }
-            uint max_DR_slot = (uint)dr_max_opnd_accessible_spill_slot();
             if (slot > max_DR_slot) {
                 /* This is not a drreg spill, but some TLS access by
                  * tool instrumentation (i#2035).
```

Comment:
So the if() on line 1639 with the big comment can be removed here?

Reply:
No.

The if (slot > max_DR_slot ) is for the "> raw tls", hence used by tool and also the "hidden slot". The hidden slot part already worked but kind of coincidentally because we wrapped around, it just makes it explicit.

The if (slot > max_DR_slot)  is to recognize the stolen scratch slot used by DR. It's not robust and can probably still be broken by a client? Would leave comment there.

Reply:
I think you meant your two if()'s to be different.  But I think I see: line 1639 is the 3rd public DR slot while this CL is looking for the non-public DR slot.

Reply:
Ack